File: Swift/BankaiCore/Tests/BankaiCoreTests/Repository/RepositoryFixtures.swift
//
//  IRepositoryFixtures.swift
//  
//
//  Created by Sergio Daniel on 1/09/24.
//

@testable import BankaiCore
import Foundation

struct TestObject: Traceable, Equatable {
    let id: Int
    let name: String
    let updatedAt: Date?
}

struct TestFilter: Filter {
    var onlyLocally: Bool
    var keywords: String?
}

class FixtureLocal: LocalDataSource {
    typealias F = TestFilter
    typealias T = TestObject
    
    var storeFixture: (_ items: [T]) async throws -> Void
    var retrieveFixture: (F?) async throws -> [T]
    
    init(
        storeFixture: @escaping (_ items: [T]) async throws -> Void,
        retrieveFixture: @escaping (F?) async throws -> [T]
    ) {
        self.storeFixture = storeFixture
        self.retrieveFixture = retrieveFixture
    }
    
    func prepare() {}
    
    func store(_ items: [T]) async throws {
        print("Stored items: \(items)")
        try await self.storeFixture(items)
    }
    
    func retrieve(filter: TestFilter?) async throws -> [T] {
        return try await retrieveFixture(filter)
    }
}

class FixtureRemote: RemoteDataSource {
    typealias F = TestFilter
    typealias T = TestObject
    
    static var name = "test-remote"
    
    var label: String
    
    var pushFixture: (_ items: [T]) async throws -> Void
    var pullFixture: (_ filter: FilterType?) async throws -> [T]
    
    init(
        label: String,
        pushFixture: @escaping (_ items: [T]) async throws -> Void,
        pullFixture: @escaping (_ filter: FilterType?) async throws -> [T]
    ) {
        self.label = label
        self.pushFixture = pushFixture
        self.pullFixture = pullFixture
    }
    
    func prepare() {}
    
    func pull(filter: TestFilter?) async throws -> [T] {
        try await Task.sleep(for: 0.5)
        return try await self.pullFixture(filter)
    }
    
    func push(_ items: [T]) async throws {
        print("Pushed items: \(items)")
        try await self.pushFixture(items)
    }
}

File: C#/BankaiCore/BankaiCoreTests/Support/RepositoryFixtures.cs
using BankaiCore.Repository;
using Microsoft.Reactive.Testing;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BankaiCoreTests.Support;

public struct TestObject: Traceable
{
    public string id { get; init; }
    public string name { get; init; }

    public DateTime? updatedAt { get; set; }
}

struct TestFilter : Filter
{
    public bool? onlyLocally { get; set; }
    public string? keywords {  get; set; }

    public static TestFilter none => new() { onlyLocally = null };
}

internal class FixtureLocal<T, F>: LocalDataSource<T, F>
    where T : Traceable
    where F : Filter
{
    /// <summary>
    /// Function holding the logic to store the fixture
    /// </summary>
    event Func<List<T>, Task> storeFixture;

    /// <summary>
    /// Function holding the logic to retrieve the fixture
    /// </summary>
    event Func<F?, Task<List<T>>> retrieveFixture;

    internal FixtureLocal(
        Func<List<T>, Task> storeFixture, 
        Func<F, Task<List<T>>> retrieveFixture
    )
    {
        this.storeFixture = storeFixture;
        this.retrieveFixture = retrieveFixture;
    }

    public Task prepare() => Task.CompletedTask;

    /// <summary>
    /// Method that calls inner logic func to retrieve items given a filter
    /// </summary>
    /// <param name="filter">The filter to base</param>
    /// <returns></returns>
    public Task<List<T>> retrieve(F? filter)
        => retrieveFixture(filter);

    /// <summary>
    /// Method that calls inner logic func to store items
    /// </summary>
    /// <param name="items"></param>
    /// <returns></returns>
    public Task store(List<T> items)
        => storeFixture(items);
}

internal class FixtureRemote<T, F>: RemoteDataSource<T, F>
    where T : Traceable
    where F : Filter
{
    public static string name { get; } = "test-remote";
    public string label;

    event Func<F?, Task<List<T>>> pullFixture;
    event Func<List<T>, Task> pushFixture;

    internal FixtureRemote(
        String label,
        Func<F?, Task<List<T>>> pullFixture,
        Func<List<T>, Task> pushFixture
    )
    {
        this.label = label;
        this.pullFixture = pullFixture;
        this.pushFixture = pushFixture;
    }

    public Task prepare() => Task.CompletedTask;

    public Task<List<T>> pull(F? filter)
        => pullFixture(filter);

    public Task push(List<T> items)
        => pushFixture(items);
}